osutil/vfs/vfs.go

```diff
@@ -0,0 +1,157 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package vfs implements a virtual file system sufficient to predict how mount,
+// bind-mount and unmount work. The [VFS.Stat] function returns a [fs.FileInfo] whose
+// [fs.FileInfo.Sys] function returns information from the underlying [fs.StatFS].
+// The [VFS.Open] function panics, allowing the implementation to be simpler.
+package vfs
+
+import (
+	"errors"
+	"io/fs"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+var (
+	errNotDir     = errors.New("not a directory")
+	errNotMounted = errors.New("not mounted")
+	errMountBusy  = errors.New("mount is busy")
+)
+
+// MountID is a 64 bit identifier of a mount.
+type MountID int64
+
+// RootMountID is the mount ID of the root file system in any VFS.
+const RootMountID MountID = -1 // LSMT_ROOT
+
+// VFS models a virtual file system.
+type VFS struct {
+	mu          sync.RWMutex
+	mounts      []*mount
+	nextMountID MountID
+}
+
+// NewVFS returns a VFS with the given root file system mounted.
+func NewVFS(rootFS fs.StatFS) *VFS {
+	return &VFS{mounts: []*mount{{
+		mountID:  RootMountID,
+		parentID: RootMountID, // The rootfs is its own parent to prevent being unmounted.
+		isDir:    true,
+		fsFS:     rootFS,
+	}}}
+}
+
+// dom returns the mount that dominates a given path.
+//
+// Out of all the mounts in the VFS, the last one that dominates a given path, wins.
+// Mounts are searched back-to-front. The search has linear complexity.
+//
+// The returned index is the index of the returned mount in [VFS.mounts].
+func (v *VFS) dom(path string) (idx int, m *mount, suffix, fsPath string) {
+	for idx = len(v.mounts) - 1; idx >= 0; idx-- {
+		m = v.mounts[idx]
+		suffix, fsPath, ok := m.isDom(path)
+		if ok {
+			return idx, m, suffix, fsPath
+		}
+	}
+
+	panic("We should have found the rootfs while looking for mount dominating " + path)
+}
+
+// mount keeps track of mounted file systems.
+type mount struct {
+	mountID    MountID
+	parentID   MountID
+	mountPoint string // Path of the mount point in the VFS. This might be a file.
+	rootDir    string // Path of fsFS that is actually mounted.
+	isDir      bool   // Mount is attached to a directory.
+	fsFS       fs.StatFS
+}
+
+// isDom returns the dominated suffix and file system path if the mount dominates the given path.
+//
+// A mount dominates the path in one of two cases:
+//
+//  1. The path is the same as the mount point.
+//  2. The mount point is a directory and the mount point subtree prefix is a prefix of
+//     the path.
+//
+// A mount point subtree prefix is the mount point followed by a directory separator
+// except for when the mount point is the empty string to denote the root directory
+// which dominates all the paths.
+func (m *mount) isDom(path string) (domSuffix, fsPath string, ok bool) {
```

Comment:
Also, I would prefer if this returned errors instead of the `ok` bool, with one of the errors be something like `ErrNotParentMount` which could be understood by `VFS.mount` to skip.